[SPARK-27996][WEBUI] Send HTTP redirect using the correct protocol when Spark History server is deployed behind the reverse proxy

ishikin opened the pull request:
### What changes were proposed in this pull request?
This PR adds method `updateProtocolFromHeader(request: HttpServletRequest, url: String): String` to JettyUtils.scala and calls this method from when Spark History Server is generating a redirect response.

### Why are the changes needed?
This change is needed to address bug SPARK-27996 in History Server. Please note that similar change is likely needed in Spark UI. The scope of this PR is only to address the problem in History Server.

### Does this PR introduce any user-facing change?
Yes, the Spark History server deployed in Kubernetes now correctly opens the job page.

### How was this patch tested?
The patch was tested by building and deploying Spark History Server to Kubernetes cluster and then observing the behavior of the History Server UI. The regression testing was not performed. I will appreciate is someone can can confirm the Spark History Server still works when deployed on regular infrastructure (e.g. not behind the reverse proxy).

Commit:
Address SPARK-27996 for History Server

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

dongjoon-hyun commented:
ok to test

dongjoon-hyun commented:
Thank you for making a PR against `master` branch.

SparkQA commented:
**[Test build #117269 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117269/testReport)** for PR 27311 at commit [`4896e19`](https://github.com/apache/spark/commit/4896e19f5e7ffd8f87cf27e266d783f8baed8bcd).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/22029/
Test PASSed.

SparkQA commented:
**[Test build #117269 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117269/testReport)** for PR 27311 at commit [`4896e19`](https://github.com/apache/spark/commit/4896e19f5e7ffd8f87cf27e266d783f8baed8bcd).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/117269/
Test FAILed.

dongjoon-hyun reviewed (changes requested):
Hi, @ishikin .
The UT failure looks relevant one. Could you take a look at `ApplicationCacheSuite`?
```
org.apache.spark.deploy.history.ApplicationCacheSuite.redirect includes query params
```

igor-shikin-c1 committed:
Fix ApplicationCacheSuite redirect unit test, add unit test for X-Forwarded-Proto header

SparkQA commented:
**[Test build #117363 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117363/testReport)** for PR 27311 at commit [`48f6bc5`](https://github.com/apache/spark/commit/48f6bc57cd1eae35e8a813deedfc99c9b2346711).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/22123/
Test PASSed.

ishikin commented:
Hi @dongjoon-hyun,
Sorry, didn't realize master has unit test coverage for History Server now. I have originally submitted this PR against 2.4 branch and there was none.
I have corrected the unit test for redirect and added unit test covering the function I am adding.
Thanks!

SparkQA commented:
**[Test build #117363 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117363/testReport)** for PR 27311 at commit [`48f6bc5`](https://github.com/apache/spark/commit/48f6bc57cd1eae35e8a813deedfc99c9b2346711).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/117363/
Test PASSed.

dongjoon-hyun commented on core/src/test/scala/org/apache/spark/deploy/history/ApplicationCacheSuite.scala:
```diff
@@ -380,4 +382,27 @@ class ApplicationCacheSuite extends SparkFunSuite with Logging with MockitoSugar
     verify(resp).sendRedirect("http://localhost:18080/history/local-123/jobs/job/?id=2")
   }
 
+  test("redirect respects X-Forwarded-Proto header") {
```
Thank you for adding this. It would be better if you follow the guideline (https://spark.apache.org/contributing.html)
```scala
- test("redirect respects X-Forwarded-Proto header") {
+ test("SPARK-27996: redirect respects X-Forwarded-Proto header") {
```

ishikin replied:
I have changed the unit test name

dongjoon-hyun commented on core/src/test/scala/org/apache/spark/deploy/history/ApplicationCacheSuite.scala:
```diff
@@ -370,7 +370,9 @@ class ApplicationCacheSuite extends SparkFunSuite with Logging with MockitoSugar
     val request = mock[HttpServletRequest]
     when(request.getMethod()).thenReturn("GET")
-    when(request.getRequestURI()).thenReturn("http://localhost:18080/history/local-123/jobs/job/")
+    when(request.getRequestURI()).thenReturn("/history/local-123/jobs/job/")
```
Ur, do we need this change? Usually, we try to avoid changing other test coverage.

ishikin replied:
getRequestURI() was incorrectly mocked here. getRequestURI() does not return protocol or host name, only part of the path after and until the query parameters.

dongjoon-hyun replied:
Oh, was it? Thanks!

dongjoon-hyun commented on core/src/test/scala/org/apache/spark/deploy/history/ApplicationCacheSuite.scala:
```diff
@@ -370,7 +370,9 @@ class ApplicationCacheSuite extends SparkFunSuite with Logging with MockitoSugar
-    when(request.getRequestURI()).thenReturn("http://localhost:18080/history/local-123/jobs/job/")
+    when(request.getRequestURI()).thenReturn("/history/local-123/jobs/job/")
+    when(request.getRequestURL())
+      .thenReturn(new StringBuffer("http://localhost:18080/history/local-123/jobs/job/"))
```
I'm wondering if we need `new StringBuffer` here.

ishikin replied:
getRequestURL() returns a StringBuffer unlike getRequestURI() that returns a plain String, so, yes, we need to provide StringBuffer type to thenReturn.

dongjoon-hyun replied:
Got it.

igor-shikin-c1 committed:
Update the unit test name

SparkQA commented:
**[Test build #117373 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117373/testReport)** for PR 27311 at commit [`d072798`](https://github.com/apache/spark/commit/d0727980968ee7a0bd3807298273bc4f94b8ddfd).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/22133/
Test PASSed.

SparkQA commented:
**[Test build #117373 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117373/testReport)** for PR 27311 at commit [`d072798`](https://github.com/apache/spark/commit/d0727980968ee7a0bd3807298273bc4f94b8ddfd).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/117373/
Test FAILed.

dongjoon-hyun commented:
Retest this please.

SparkQA commented:
**[Test build #117390 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117390/testReport)** for PR 27311 at commit [`d072798`](https://github.com/apache/spark/commit/d0727980968ee7a0bd3807298273bc4f94b8ddfd).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/22150/
Test PASSed.

SparkQA commented:
**[Test build #117390 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117390/testReport)** for PR 27311 at commit [`d072798`](https://github.com/apache/spark/commit/d0727980968ee7a0bd3807298273bc4f94b8ddfd).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/117390/
Test PASSed.

ishikin commented:
Hi @dongjoon-hyun,
Not sure what is the next step since this is my first PR for Spark. 
Please let me know if anything else is needed from my side.
Thanks!

dongjoon-hyun commented:
@ishikin . Sorry. I also forgot this PR due to the other community work items(`2.4.5 RC2` preparation and some correctness issues). I'll revisit this PR as soon as possible.
BTW, if there is a delay on your PRs, you can ping another committer, too. Every committer can review and merge.

dongjoon-hyun commented:
BTW, the following might be a risk at the other Spark resource managers, especially, YARN/MESOS.
> The regression testing was not performed. I will appreciate is someone can can confirm the Spark History Server still works when deployed on regular infrastructure (e.g. not behind the reverse proxy).
